packages/xl-ai/src/streamTool/callLLMWithStreamTools.ts

```diff
@@ -0,0 +1,283 @@
+    _streamObjectOptions?: Partial<
+      Parameters<typeof streamObject<{ operations: any[] }>>[0]
+    >;
+  },
+  onStart: () => void = () => {
+    // noop
+  },
+): Promise<OperationsResult<T>> {
+  const { _streamObjectOptions, ...rest } = opts;
+
+  if (
+    _streamObjectOptions &&
+    ("output" in _streamObjectOptions || "schema" in _streamObjectOptions)
+  ) {
+    throw new Error("Cannot provide output or schema in _streamObjectOptions");
+  }
+
+  const schema = jsonSchema(createStreamToolsArraySchema(streamTools));
+
+  const options = {
+    // non-overridable options for streamObject
+    output: "object" as const,
+    schema,
+    // configurable options for streamObject
+
+    // - optional, with defaults
+    // mistral somehow needs "auto", while groq/llama needs "tool"
+    // TODO: further research this and / or make configurable
+    // for now stick to "tool" by default as this has been tested mostly
+    mode: rest.model.provider === "mistral.chat" ? "auto" : "tool",
+    maxRetries: 2,
+    //  - mandatory ones:
+    ...rest,
+
+    // extra options for streamObject
+    ...((opts._streamObjectOptions ?? {}) as any),
+  };
+
+  const ret = streamObject<{ operations: any }>(options);
+
+  let _operationsSource: OperationsResult<T>["operationsSource"];
+
+  return {
+    streamObjectResult: ret,
+    generateObjectResult: undefined,
+    get operationsSource() {
+      if (!_operationsSource) {
+        _operationsSource = createAsyncIterableStreamFromAsyncIterable(
+          preprocessOperationsStreaming(
+            filterNewOrUpdatedOperations(
+              streamOnStartCallback(
+                partialObjectStreamThrowError(ret.fullStream),
+                onStart,
+              ),
+            ),
+            streamTools,
+          ),
+        );
+      }
+      return _operationsSource;
+    },
+  };
+}
+
+async function* streamOnStartCallback<T>(
+  stream: AsyncIterable<T>,
+  onStart: () => void,
+): AsyncIterable<T> {
+  let first = true;
+  for await (const chunk of stream) {
+    if (first) {
+      onStart();
+      first = false;
+    }
+    yield chunk;
+  }
+}
+
+// adapted from https://github.com/vercel/ai/blob/5d4610634f119dc394d36adcba200a06f850209e/packages/ai/core/generate-object/stream-object.ts#L1041C7-L1066C1
+// change made to throw errors (with the original they're silently ignored)
+function partialObjectStreamThrowError<PARTIAL>(
+  stream: AsyncIterableStream<ObjectStreamPart<PARTIAL>>,
+): AsyncIterableStream<PARTIAL> {
+  return createAsyncIterableStream(
+    stream.pipeThrough(
+      new TransformStream<ObjectStreamPart<PARTIAL>, PARTIAL>({
+        transform(chunk, controller) {
+          switch (chunk.type) {
+            case "object":
+              controller.enqueue(chunk.object);
+              break;
+
+            case "text-delta":
+            case "finish":
+              break;
+            case "error":
+              throw chunk.error;
```

Comment:
Why throw on error instead of using the `controller.error` https://developer.mozilla.org/en-US/docs/Web/API/TransformStreamDefaultController/error

Reply:
oof not sure why I didn't do so! will test and apply, thanks